camel/tasks/task.py

```diff
@@ -0,0 +1,257 @@
+        self.state = "OPEN"
+        self.result = ""
+
+    def update_task_result(self, message: BaseMessage):
+        self.result = message.content
+        self.state = "DONE"
+
+    def set_id(self, id: str):
+        self.id = id
+
+    def set_state(self, state: Literal["OPEN", "RUNNING", "DONE", "DELETED"]):
+        assert state in TASK_STATES, f"Invalid state `{state}`"
+
+        self.state = state
+        if state == "DONE":
+            for subtask in self.subtasks:
+                if subtask.state != "DELETED":
+                    subtask.set_state(state)
+        elif state == "RUNNING":
+            if self.parent is not None:
+                self.parent.set_state(state)
+
+    def add_subtask(self, task: "Task"):
+        task.parent = self
+        self.subtasks.append(task)
+
+    def get_running_task(self) -> Optional["Task"]:
+        for sub in self.subtasks:
+            if sub.state == "RUNNING":
+                return sub.get_running_task()
+        if self.state == "RUNNING":
+            return self
+        return None
+
+    def to_string(self, indent="") -> str:
+        _str = f"{indent}Task {self.id}: {self.content}\n"
+        for subtask in self.subtasks:
+            _str += subtask.to_string(indent + "  ")
+        return _str
+
+
+def parse_response(response: str, task_id: Optional[str]) -> List[Task]:
+    pattern = "<task>(.*?)</task>"
+    tasks_content = re.findall(pattern, response, re.DOTALL)
+
+    tasks = []
+    if task_id is None:
+        task_id = "0"
+    for i, content in enumerate(tasks_content):
+        tasks.append(Task(content=content.strip(), id=f"{task_id}.{i}"))
+    return tasks
+
+
+class TaskManager:
+    """
+    TaskManager is used to manage tasks.
+
+    Attributes:
+        root_task: The root task.
+        tasks: The ordered tasks.
+        task_map: A map for task.id to Task.
+        current_task_id: The current "RUNNING" task.id.
+    """
+
+    def __init__(self, task: Task):
+        self.root_task: Task = task
+        self.current_task_id: str = task.id
+        self.tasks: List[Task] = [task]
+        self.task_map: Dict[str, Task] = {task.id: task}
+
+    def is_taskid_existed(self, task_id: str) -> bool:
+        return task_id in self.task_map
+
+    @property
+    def current_task(self) -> Optional[Task]:
+        return self.task_map.get(self.current_task_id, None)
+
+    @staticmethod
+    def _topological_sort(tasks: List[Task]) -> List[Task]:
+        stack = []
+        visited = set()
+
+        # recursive visit the vertices
+        def visit(task: Task):
+            if task.id in visited:
+                return
+            visited.add(task.id)
+
+            # go deep for dependencies
+            for sub_task in task.subtasks:
+                visit(sub_task)
+
+            # add current task to stack which have no dependencies.
+            stack.append(task)
+
+        for task in tasks:
+            visit(task)
```

Comment:
How about the dependency between the subtasks. I mean that there may be task 0.0 should be executed before task 0.1.

Reply:
Maoquan said that it depend on the prompt of decompose task. Maybe we can treat all subtasks as if they were performed in sequence first.